### PR TITLE
WIP issue 56 add is-unauthenticated and unauthenticated event

### LINKED ISSUE #54: hoodie.account.fetch(), .update(), .profile.fetch() and .profile.update() should fail with UnauthenticatedError if session is invalid, but currently fails with "Not FoundError"
To simulate an unauthenticated state, replace `session.id` in `localStorage.account` with a bogus value, then reload the page and do `account.fetch()`

---

- [x] claim this issue (@realtin)
- [x] start a PR (#62)
- [x] add unit and integration tests for `account.fetch()`. In the latter, make sure that an "unauthenticate" event gets triggered
- [ ] fix account.fetch()
- [ ] repeat steps for `account.update`, `account.profile.fetch` and `account.profile.update`
- [ ] review & merge
- [ ] update https://github.com/gr2m/milestones/issues/88

### LINKED ISSUE #56: Implement account.hasInvalidSession()
If a session gets invalidated, "unauthenticate" events get triggered, for example when `account.fetch()` is called. Once unauthenticated, `account.isUnauthenticated` shall return true, otherwise false. The unauthenticate state shall also be persisted in localStorage.

`account.signIn` with the same username creates a new valid session and emits the `reauthenticate` event

---

- [x] claim this issue (comment below or assign yourself) (@toh82)
- [x] start a PR (#72)
- [x] add unit tests for `account.hasInvalidSession()`
- [x] implement `account.hasInvalidSession()`, and persist the unauthenticate state in localStorage
- [ ] update [README](https://github.com/hoodiehq/hoodie-client-account) (add `account.hasInvalidSession` and remove issue reference in Events)
- [ ] review & merge
- [ ] update https://github.com/gr2m/milestones/issues/88

### LINKED ISSUE #64: use `state.url` instead of `state.baseUrl` in fetch.js and profile-fetch.js
Reported in: https://github.com/hoodiehq/hoodie-client-account/issues/61#
When reproducing the Error from #54 a GET Request to `http://127.0.0.1:8080/undefined/session/account` is send. 

Fix that by using `state.url` instead of `state.baseUrl` 

-------
- [x] claim this issue (@Endi1)
- [x] start a PR
- [x] uncomment integration test `test/integration/fetch-url-test.js`
- [x] fix `state.baseUrl` in fetch.js
- [x] fix `state.baseUrl` in profile-fetch.js
- [x] search for more appearance of `state.baseUrl`
- [x] run tests and make sure that all are green
- [ ] review & merge
- [ ] check if https://github.com/hoodiehq/hoodie-client-account/pull/62 is resolved

### PR DESCRIPTION
**Pull request by toh82**
closes #54
closes #56 
closes #64

### CODE REVIEW

**Comment by toh82** [comment]
i'm not sure about the implementation of `isUnauthenticated` if it is really only about checking if session node exists the `is-signed-in` does that job already. 

And i'm thinking of to put the event trigger for `unauthenticated` in something what is used accross the authentication methods like "sign-out, fetch" to not add the event emitter in every module, but yet donno where.

**Comment by gr2m** [comment]
You are unauthenticated of you Are signed in, but your session is invalid

**Comment by toh82** [comment]
@gr2m what means invalid in that case, no `session.id` given, because `session` object seems to exist in any case? Or should `isUnauthenticated` use `account.validate` with a new written validator, or a request for validating the session?

**Comment by gr2m** [comment]
yeah I agree this is confusing. Here is an example of what I mean

1. you sign in on your phone
2. you sign in on your laptop and change your password
3. on your phone, you do `account.fetch()`, but it will fail with `401` and trigger an `unauthentciate` error.

I thought that the `unauthenticate` event would be enough, and maybe it is, I’m still not 100% sure. Problem is that when you load the page initially, you might want to show the user right away that they do not have a valid session and need to sign in again, and not make them wait until they do `.fetch()` again. So the `.isUnauthenticated()` method can be called to check that during rendering.

But how about we rename it to `.hasInvalidSession()`. Would that be more clear?

**Comment by toh82** [comment]
Still todo:
- invalid state should be written to local storage
- update should emit unauthenticate and set invalid state
- if state in locale storeage is already invalid then `.hasInvalidSession()` should return true
- unauthenticate event should only be emitted once

**Comment by gr2m** [comment]
@toh82 what’s the state here? Are you still working on it?

**Comment by toh82** [comment]
@gr2m i updated all the things \o/ and fixed an [interesting bug about profile fetch](https://github.com/hoodiehq/hoodie-client-account/commit/a537f940b5adf2469f15f0b333323d490288a3ce) meanwhile. Would be awesome if you could have a look, afterwards I'll squash it further.

I'm still sad about the code duplication in the tests, anyways, if youre fine i could make the 401 response as a utility to avoid duplication across the fetch and update modules.

File: test/integration/fetch-url-test.js [hunk 1]
```diff
@@ -40,4 +40,4 @@ test('account.fetch() and account.profil.fetch()', function (t) {
 
   .catch(t.fail)
 })
-*/
\ No newline at end of file
+*/
```
Comment (gr2m) [inline test/integration/fetch-url-test.js]: I think this PR fixes https://github.com/hoodiehq/hoodie-client-account/issues/64, right? So we can uncomment this while at it?

**Comment by gr2m** [comment]
_fantastic_ work :) If you could look into my minor comments and then clean up the commits, this is good to merge :+1:

**Comment by gr2m** [comment]
ah, forgot one thing, please update the README.md file, too. Add the `account.hasInvalidSession()` method and remove the `#56` reference in the `unauthenticate` event

**Comment by toh82** [comment]
activated `integration/fetch-url-test.js` and searched for more `.baseUrl` but found nothing, so this closes #64 too. Which commit did you meant with:

> fyi that is not a feat: ..., it’s a refactor: ... as no logic is changed, just internal implementation

**Pull request merged**